esapi/esapi_util/esys_mu.c

```diff
@@ -0,0 +1,1200 @@
+        buffer, size, offset, dst);
+    size_t offset_loc;
+    IESYSC_PARAM_DECRYPT dst_loc;
+    TSS2_RC ret = Tss2_MU_UINT32_Unmarshal(buffer, size,
+        &offset_loc, &dst_loc);
+    if (ret != TSS2_RC_SUCCESS) {
+        LOG_ERROR("Unmarshaling the base type");
+        return ret;
+    }
+    ret = Tss2_MU_IESYSC_PARAM_DECRYPT_check(&dst_loc);
+    if (ret != TSS2_RC_SUCCESS) {
+        LOG_ERROR("Bad value %"PRIx32 "", dst_loc);
+        return ret;
+    }
+    if (offset != NULL)
+        *offset = offset_loc;
+    if (dst != NULL)
+        *dst = dst_loc;
+    LOG_TRACE("return: dst=%p value=%"PRIx32 "", dst, dst_loc);
+    return TSS2_RC_SUCCESS;
+}
+
+/**
+ * Check, if a variable has a possible value of type IESYSC_PARAM_DECRYPT.
+ * @param[in] in variable to check.
+ * @retval TSS2_RC_SUCCESS on success.
+ */
+TSS2_RC
+Tss2_MU_IESYSC_PARAM_DECRYPT_check(
+    const IESYSC_PARAM_DECRYPT *in)
+{
+    LOG_TRACE("called: in=%p", in);
+    if (in == NULL) {
+        LOG_ERROR("in==NULL");
+        return TSS2_SYS_RC_BAD_REFERENCE;
+    }
+    /* No Error-Messages, since this function may fail for a good reasons. */
+    if (FALSE
+        || (*in == DECRYPT)
+        || (*in == NO_DECRYPT)) {
+        return TSS2_RC_SUCCESS;
+    } else {
+        return TSS2_SYS_RC_BAD_VALUE;
+    }
+    return TSS2_RC_SUCCESS;
+}
+/**
+ * Marshal a constant of type IESYSC_TYPE_POLICY_AUTH into a byte buffer.
+ * @param[in] src constant to be marshalled.
+ * @param[in,out] buffer Buffer to write result into (may be NULL)
+ * @param[in] size Size of the buffer.
+ * @param[in,out] offset Offset inside the buffer (may be NULL.
+ * @retval TSS2_RC_SUCCESS on success.
+ * @retval TSS2_ESYS_RC_BAD_REFERENCE if src==NULL.
+ * @retval TSS2_ESYS_RC_INSUFFICIENT_BUFFER if remaining buffer is insufficient.
+ */
+TSS2_RC
+Tss2_MU_IESYSC_TYPE_POLICY_AUTH_Marshal(
+    const IESYSC_TYPE_POLICY_AUTH src,
+    uint8_t *buffer,
+    size_t size,
+    size_t *offset)
+{
+    LOG_TRACE("called: src=%"PRIx32 " buffer=%p size=%zu offset=%p", src,
+        buffer, size, offset);
+    return Tss2_MU_UINT32_Marshal(src, buffer, size, offset);
+}
+
+/**
+ * Unmarshal a constant of type IESYSC_TYPE_POLICY_AUTH from a byte buffer.
+ * @param[in,out] buffer Buffer to read data from.
+ * @param[in] size Size of the buffer.
+ * @param[in,out] offset Offset inside the buffer
+ *                (being updated during marshalling).
+ * @param[out] dst variable to store the result in.
+ * @retval TSS2_RC_SUCCESS on success.
+ */
+TSS2_RC
+Tss2_MU_IESYSC_TYPE_POLICY_AUTH_Unmarshal(
+    const uint8_t *buffer,
+    size_t size,
+    size_t *offset,
+    IESYSC_TYPE_POLICY_AUTH *dst)
+{
+    LOG_TRACE("called: buffer=%p size=%zu offset=%p dst=%p",
+        buffer, size, offset, dst);
+    size_t offset_loc;
+    IESYSC_TYPE_POLICY_AUTH dst_loc;
+    TSS2_RC ret = Tss2_MU_UINT32_Unmarshal(buffer, size,
+        &offset_loc, &dst_loc);
```

Comment:
`offset_loc` passed to `Tss2_MU_UINT32_Unmarshal` uninitialized. If the intent is to unmarshal the type from the provided buffer at the offset provided in the parameter `offset` then it looks like we may end up starting from an undefined offset.